src/main/scala/scorex/core/network/NodeViewSynchronizer.scala

```diff
@@ -214,65 +263,135 @@ class NodeViewSynchronizer[TX <: Transaction,
+          val parsed: Iterable[PMOD] = parseModifiers(requestedModifiers, serializer, remote)
+          parsed.foreach(pmod => putToCacheIfValid(remote, pmod))
+          if (parsed.nonEmpty) {
+            // remove elements from cache if it's size exceeds the limit, reset status for removed modifiers
+            val cleared = modifiersCache.cleanOverfull()
+            if (cleared.nonEmpty) {
+              log.debug(s"${cleared.size} modifiers ${idsToString(typeId, cleared.map(_.id))} were removed from cache")
+              cleared.foreach(removed => deliveryTracker.stopProcessing(removed.id))
+            }
+            // send changed cache to node view holder to try to apply new modifiers
+            viewHolderRef ! ChangedCache[PMOD, HR, ModifiersCache[PMOD, HR]](modifiersCache)
+          }
+        case _ =>
+          log.error(s"Undefined serializer for modifier of type $typeId")
       }
   }
 
-  //local node sending object ids to remote
-  protected def requestFromLocal: Receive = {
-    case RequestFromLocal(peer, modifierTypeId, modifierIds) =>
+  /**
+    * Put `pmod` to `modifiersCache` if `pmod` is valid or may be valid in the future,
+    * penalize peer if `pmod` is permanently invalid
+    */
+  @SuppressWarnings(Array("org.wartremover.warts.IsInstanceOf"))
+  private def putToCacheIfValid(remote: ConnectedPeer, pmod: PMOD): Unit = {
+    historyReaderOpt match {
+      case Some(hr) =>
+        hr.applicableTry(pmod) match {
+          case Failure(e) if e.isInstanceOf[MalformedModifierError] =>
+            log.warn(s"Modifier ${pmod.encodedId} is permanently invalid", e)
+            deliveryTracker.onInvalid(pmod.id)
+            penalizeMisbehavingPeer(remote)
+          case _ =>
+            modifiersCache.put(pmod.id, pmod)
+        }
+      case None =>
+        log.error("Got modifier while history reader is not ready")
+        modifiersCache.put(pmod.id, pmod)
+    }
+  }
 
-      if (modifierIds.nonEmpty) {
-        val msg = Message(requestModifierSpec, Right(modifierTypeId -> modifierIds), None)
-        peer.handlerRef ! msg
+  /**
+    * Parse modifiers using specified serializer, check that it's id is equal to the declared one,
+    * penalize misbehaving peer for every incorrect modifier,
+    * call deliveryTracker.onReceive() for every correct modifier to update it's status
+    *
+    * @return collection of parsed modifiers
+    */
+  private def parseModifiers[M <: NodeViewModifier](modifiers: Map[ModifierId, Array[Byte]],
+                                                    serializer: Serializer[M],
+                                                    remote: ConnectedPeer): Iterable[M] = {
+    modifiers.flatMap { case (id, bytes) =>
+      serializer.parseBytes(bytes) match {
+        case Success(mod) if id == mod.id =>
+          deliveryTracker.onReceive(id)
+          Some(mod)
+        case _ =>
+          penalizeMisbehavingPeer(remote)
```

Comment:
Why status is not changed from requested? So you can e.g. ban a peer as a result of "penalize" but still wait from it?

Reply:
Added a comment